File: src/FXRateDashboard/Utilities/WindowPlacementHelper.cs
using Point = System.Windows.Point;
using Rect = System.Windows.Rect;
using Size = System.Windows.Size;

namespace FXRateDashboard.Utilities;

public static class WindowPlacementHelper
{
    private const double EdgeAnchorThreshold = 24;

    public static Point CalculateClampedTopLeft(Rect workArea, Rect currentBounds, Size targetSize)
    {
        var isAnchoredToRight = workArea.Right - currentBounds.Right <= EdgeAnchorThreshold;
        var isAnchoredToBottom = workArea.Bottom - currentBounds.Bottom <= EdgeAnchorThreshold;

        var targetLeft = isAnchoredToRight
            ? currentBounds.Right - targetSize.Width
            : currentBounds.Left;
        var targetTop = isAnchoredToBottom
            ? currentBounds.Bottom - targetSize.Height
            : currentBounds.Top;

        if (targetLeft + targetSize.Width > workArea.Right)
        {
            targetLeft = workArea.Right - targetSize.Width;
        }

        if (targetTop + targetSize.Height > workArea.Bottom)
        {
            targetTop = workArea.Bottom - targetSize.Height;
        }

        targetLeft = Math.Max(workArea.Left, targetLeft);
        targetTop = Math.Max(workArea.Top, targetTop);

        return new Point(targetLeft, targetTop);
    }
}

File: macos/FXRateDashboard/Sources/FXRateDashboardKit/Support/WindowPlacementHelper.swift
import CoreGraphics

public enum WindowPlacementHelper {
    private static let edgeAnchorThreshold: CGFloat = 24

    public static func calculateClampedTopLeft(
        workArea: CGRect,
        currentBounds: CGRect,
        targetSize: CGSize
    ) -> CGPoint {
        let isAnchoredToRight = workArea.maxX - currentBounds.maxX <= edgeAnchorThreshold
        let isAnchoredToBottom = workArea.maxY - currentBounds.maxY <= edgeAnchorThreshold

        var targetLeft = isAnchoredToRight ? currentBounds.maxX - targetSize.width : currentBounds.minX
        var targetTop = isAnchoredToBottom ? currentBounds.maxY - targetSize.height : currentBounds.minY

        if targetLeft + targetSize.width > workArea.maxX {
            targetLeft = workArea.maxX - targetSize.width
        }

        if targetTop + targetSize.height > workArea.maxY {
            targetTop = workArea.maxY - targetSize.height
        }

        targetLeft = max(workArea.minX, targetLeft)
        targetTop = max(workArea.minY, targetTop)

        return CGPoint(x: targetLeft, y: targetTop)
    }
}
